lib/xapi-stdext-unix/unixext.ml

```diff
@@ -34,6 +34,27 @@ let mkdir_rec dir perm =
     mkdir_safe dir perm in
   p_mkdir dir
 
+(** removes a file or recursively removes files/directories below a directory without following
+    symbolic links. If path is a directory, it is only removed if rm_top is true *)
+let rm_rec ?(rm_top = true) path =
+  let ( // ) = Filename.concat in
+  let rec rm rm_top path =
+    match Unix.lstat path with
+    | exception Unix.Unix_error (Unix.ENOENT, _, _) ->
+        () (*noop*)
+    | exception e ->
+        raise e
+    | st -> (
+      match st.Unix.st_kind with
+      | Unix.S_DIR ->
+          Sys.readdir path |> Array.iter (fun file -> rm true (path // file)) ;
+          if rm_top then Unix.rmdir path
```

Comment:
I find it surprising that the `rm_top` case is handled in the recursive part and not at the top level where `rm` is called. I struggled to understand what it does in the recursive. I tested the code and it seems to do the right thing, though.

Reply:
Hmm I'm not sure how you could move rm_top outside the recursive part. The Unix.rmdir has to happen in the recursive part (and after its contents has been rm'd) and so rm_top is just a guard to stop the first iteration from doing Unix.rmdir if rm_top=false. Thereafter, rm_top=true in all the recursive calls to rm.